tests/unit/test_network_of_populations/test_visualisation.py

```diff
@@ -0,0 +1,206 @@
+        {"time": 0, "node": "hb4", "state": "S", "total": 10.0},
+        {"time": 0, "node": "hb5", "state": "S", "total": 10.0},
+        {"time": 0, "node": "hb6", "state": "S", "total": 10.0},
+        {"time": 0, "node": "hb7", "state": "S", "total": 10.0},
+        {"time": 1, "node": "hb1", "state": "S", "total": 10.0},
+        {"time": 1, "node": "hb2", "state": "S", "total": 5.0},
+        {"time": 1, "node": "hb3", "state": "S", "total": 5.0},
+        {"time": 1, "node": "hb3", "state": "E", "total": 15.0},
+        {"time": 1, "node": "hb4", "state": "S", "total": 0.0},
+        {"time": 1, "node": "hb5", "state": "S", "total": 10.0},
+        {"time": 1, "node": "hb6", "state": "S", "total": 10.0},
+        {"time": 1, "node": "hb7", "state": "S", "total": 10.0},
+    ])
+    compare_mpl_plots(vis.plot_nodes(pd.DataFrame(simple)))
+
+
+def test_plotStates_two_rows():
+    simple = pd.DataFrame([
+        {"time": 0, "node": "hb1", "state": "S", "total": 15.0},
+        {"time": 0, "node": "hb2", "state": "S", "total": 21.0},
+        {"time": 0, "node": "hb3", "state": "S", "total": 20.0},
+        {"time": 0, "node": "hb3", "state": "E", "total": 0.0},
+        {"time": 0, "node": "hb4", "state": "S", "total": 10.0},
+        {"time": 1, "node": "hb1", "state": "S", "total": 10.0},
+        {"time": 1, "node": "hb2", "state": "S", "total": 5.0},
+        {"time": 1, "node": "hb3", "state": "S", "total": 5.0},
+        {"time": 1, "node": "hb3", "state": "E", "total": 15.0},
+        {"time": 1, "node": "hb4", "state": "S", "total": 0.0},
+    ])
+    compare_mpl_plots(vis.plot_nodes(pd.DataFrame(simple)))
+
+
+def test_plotStates_single_row():
+    simple = pd.DataFrame([
+        {"time": 0, "node": "hb1", "state": "S", "total": 15.0},
+        {"time": 0, "node": "hb2", "state": "S", "total": 21.0},
+        {"time": 1, "node": "hb1", "state": "S", "total": 10.0},
+        {"time": 1, "node": "hb2", "state": "S", "total": 5.0},
+    ])
+    compare_mpl_plots(vis.plot_nodes(pd.DataFrame(simple)))
+
+
+def test_plotStates_empty_node():
+    simple = pd.DataFrame([{"time": 0, "node": "hb1", "state": "S", "total": 15.0}])
+    with pytest.raises(ValueError):
+        vis.plot_nodes(pd.DataFrame(simple), nodes=[])
+
+
+def test_plotStates_empty_states():
+    simple = pd.DataFrame([{"time": 0, "node": "hb1", "state": "S", "total": 15.0}])
+    with pytest.raises(ValueError):
+        vis.plot_nodes(pd.DataFrame(simple), states=[])
+
+
+def test_plotStates_empty_missing_column():
+    simple = pd.DataFrame([{"node": "hb1", "state": "S", "total": 15.0}])
+    with pytest.raises(ValueError):
+        vis.plot_nodes(pd.DataFrame(simple), states=[])
+
+
+def test_read_output(tmpdir):
+    df = pd.DataFrame([{"a": 10, "b": 20}])
+    df.to_csv(str(tmpdir / Path("simple.csv")), index=False)
+    path = str(tmpdir / Path("access.yaml"))
+    with open(path, "w") as fp:
+        fp.write(f"""
+data_directory: {str(tmpdir)}
```

Comment:
If you can think of a way to modify the file API to make this kind of testing easier do let me know.

Reply:
Maybe you could have some testing module with a fixture extending tmp_path behaviour so I could call `tmp_data.access_log.io.append({...})` and things like that it could make testing easier. But I haven't really thought out how feasiable that'd be.

Reply:
Hmm yeah, once the standard API implementation is done I'll think about that.